book/src/how/r1cs_constraints.md

```diff
@@ -2,49 +2,72 @@
+$Bz(r_x) = \sum_y B(r_x, y) * z(y)$
 
-#### Constraint system 
+$Cz(r_x) = \sum_y C(r_x, y) * z(y)$
 
-The constraints for a CPU step are detailed in the `get_jolt_matrices()` function in [`constraints.rs`](https://github.com/a16z/jolt/blob/main/jolt-core/src/r1cs/constraints.rs). 
+We could run a sumcheck for each of these but it's more efficient from both a prover and verifier perspective to combine into a single sumcheck by pulling a random variable $r_c$ from the Fiat-Shamir transcript and combining the linear terms.
+
+$ABCz(r_x) = \widetilde{Az}(r_x) + r_c * \widetilde{Bz}(r_x)  + r_c^2 * \widetilde{Cz}(r_x) = \sum_y{[ A(r_x, y) + r_c * B(r_x, y) + r_c^2 * C(r_x, y] ) * \widetilde{Z}(y)}$
+
+Outputs a claimed evaluation of $A(r_x, r_y), B(r_x, r_y), C(r_x, r_y), Z(r_y)$, specifically in the form $A(r_x,r_y) + r_c * B(r_x, r_y) + r_c^2 * C(r_x, r_y)$.
+
+The verifier does have a representation of these is able to evaluate without assistance.
+
+## Uniform Spartan
+Jolt's Spartan instance is highly patterned. We run the same R1CS for each cycle of the RISC-V program. We can leverage this property to increase prover and verifier efficiency.
+
+Think of a simple constraint system with the following inputs: $Q, R, S$. Each of these inputs is repeated $nc = num\_cycles$ times. 
+
+The witness vector, $Z$ now looks like $[Q_0, ... Q_{nc}, R_0, ... R_{nc}, S_{0}, ... S_{nc}, 1]$. The one in the final slot allows constants in the constraints.
+
+> [!NOTE]
+> Syntax: For bitvector $x=x_0, x_1, ... x_n$, $x_0$ represents the most-significant bit, and $x_n$ represents the least significant bit. By default sumchecks will bind from the "top", where they bind the MSB first. For some sumchecks, notably sparse ones we'll bind from the "bottom" for algorithmic efficiency reasons. 
+
+We want a property where we can easily index the Z's $cycle$ and $variable$ separately. To do so we'll pad the number of cycles to a power of two and the number of variables to a power of two.
+
+Concretely for $nc=3$ the $Z$ vector looks like $[Q_0, Q_1, Q_2, 0, R_0, R_1, R_2, 0, S_0, S_1, S_2, 0, 0, 0, 0, 0, 1]$.
+
+Now we can index by $cycle$ and $variable$ independently.
+- $Z[v_0, v_1, v_2, c_0, c_1]$ 
+- $v_0, v_1, v_2$ index the variable
+- $c_0, c_1$ index the cycle
+- $Z[0, 0, 1, c_0, c_1]$ allows you to index into all of R's values using $c_0, c_1$
+- $Z[v_0, v_1, v_2, 1, 0]$ allows you to index $Q_2, R_2, S_2$ using $v_0, v_1, v_2$ 
+
+Note that Z is also padded to a power of two, so the full Z vector is $[Q_0, Q_1, Q_2, 0, R_0, R_1, R_2, 0, S_0, S_1, S_2, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]$ 
+
+$v_0$ indexes constant vs non-constant columns.
+
+The $A, B, C$ matrices follow a similar pattern where the constraints are repeated $nc$ times and the variables are repeated $nc$ times. For example, given a 2 constraint system over the variables $Q, R, S$.
+![Big A](../imgs/big_A_r1cs.png)
```

Comment:
maybe add a legend for what the different colors mean. assuming yellow = padding, then the four columns before C should also be yellow, right?